src/main/java/org/javarosa/xpath/expr/XPathSortFunc.java

```diff
@@ -0,0 +1,146 @@
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.crypto.dsig.spec.XPathType;
+
+/**
+ * Created by amstone326 on 6/28/17.
+ */
+
+public class XPathSortFunc extends XPathFuncExpr {
+
+    public static final String NAME = "sort";
+
+    // since we accept 1-3 arguments
+    private static final int EXPECTED_ARG_COUNT = -1;
+
+    public XPathSortFunc() {
+        name = NAME;
+        expectedArgCount = EXPECTED_ARG_COUNT;
+    }
+
+    public XPathSortFunc(XPathExpression[] args) throws XPathSyntaxException {
+        super(NAME, args, EXPECTED_ARG_COUNT, true);
+    }
+
+    @Override
+    protected void validateArgCount() throws XPathSyntaxException {
+        if (args.length < 1 || args.length > 3) {
+            throw new XPathArityException(name, "between 1 and 3 arguments", args.length);
+        }
+    }
+
+    @Override
+    protected Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        List<String> sortedList;
+        if (evaluatedArgs.length == 1) {
+            sortedList = sortSingleList(FunctionUtils.toString(evaluatedArgs[0]), true);
+        } else if (evaluatedArgs.length == 2) {
+            Object lastArg = evaluatedArgs[1];
+            if (lastArg instanceof Boolean) {
+                sortedList =
+                        sortSingleList(FunctionUtils.toString(evaluatedArgs[0]),
+                                FunctionUtils.toBoolean(evaluatedArgs[1]));
+            } else {
+                sortedList = sortListByOtherList(FunctionUtils.toString(evaluatedArgs[0]),
+                        FunctionUtils.toString(evaluatedArgs[1]), true);
+            }
+        } else {
+            sortedList =
+                    sortListByOtherList(FunctionUtils.toString(evaluatedArgs[0]),
+                            FunctionUtils.toString(evaluatedArgs[1]),
+                            FunctionUtils.toBoolean(evaluatedArgs[2]));
+        }
+        return listToString(sortedList);
+    }
+
+    private List<String> sortSingleList(String spaceSeparatedString, boolean ascending) {
+        List<String> items = stringToList(spaceSeparatedString);
+        sortSingleList(items, ascending);
+        return items;
+    }
+
+    private void sortSingleList(List<String> items, final boolean ascending) {
+        Collections.sort(items, new Comparator<String>() {
+            @Override
+            public int compare(String s1, String s2) {
+                return (ascending ? 1 : -1) * s1.compareTo(s2);
+            }
+        });
+    }
+
+    private List<String> sortListByOtherList(String s1, String s2, boolean ascending) {
+        List<String> targetListItems = stringToList(s1);
+        List<String> comparisonListItems = stringToList(s2);
+
+        if (targetListItems.size() != comparisonListItems.size()) {
+            throw new XPathTypeMismatchException("Length of lists passed to sort() must match");
```

Comment:
It would possibly be worth including the list contents here, since we already have the strings. Or describing the lengths. Generally providing any cue that would let someone track why things did/didn't map.